File: contracts/api/protocol-java/src/test/java/app/epistola/protocol/PollBackoffTest.java
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.protocol;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class PollBackoffTest {

    private final PollBackoff backoff = PollBackoff.of(1_000, 30_000, 3.0);

    @Test
    void an_idle_poll_after_a_has_more_burst_returns_to_the_floor_not_to_zero() {
        // The bug every client shipped: afterHasMore() is 0 so the next poll is immediate, and
        // 0 * multiplier is still 0 — without the floor the caller issues its next request with no
        // delay, forever, as soon as a burst drains.
        assertEquals(0L, backoff.afterHasMore());
        assertEquals(1_000L, backoff.afterIdlePoll(backoff.afterHasMore()));
    }

    @Test
    void idle_polls_grow_the_interval_and_stop_at_the_maximum() {
        assertEquals(3_000L, backoff.afterIdlePoll(1_000));
        assertEquals(9_000L, backoff.afterIdlePoll(3_000));
        assertEquals(27_000L, backoff.afterIdlePoll(9_000));
        assertEquals(30_000L, backoff.afterIdlePoll(27_000));
        assertEquals(30_000L, backoff.afterIdlePoll(30_000), "capped, not growing forever");
    }

    @Test
    void results_reset_the_interval_to_the_floor() {
        assertEquals(1_000L, backoff.afterResults());
        assertEquals(1_000L, backoff.initial());
    }

    @Test
    void a_kick_only_shortens_an_interval_that_is_already_longer() {
        assertEquals(3_000L, backoff.afterKick(30_000, 3_000), "idle: shorten to the kick interval");
        assertEquals(1_000L, backoff.afterKick(1_000, 3_000), "already faster: leave it alone");
        assertEquals(3_000L, backoff.afterKick(3_000, 3_000), "exactly at it: leave it alone");
    }

    @Test
    void the_configuration_is_validated_where_it_is_supplied() {
        assertThrows(IllegalArgumentException.class, () -> PollBackoff.of(0, 30_000, 3.0));
        assertThrows(IllegalArgumentException.class, () -> PollBackoff.of(-1, 30_000, 3.0));
        assertThrows(IllegalArgumentException.class, () -> PollBackoff.of(30_000, 1_000, 3.0));
        assertThrows(IllegalArgumentException.class, () -> PollBackoff.of(1_000, 30_000, 1.0));
    }
}

File: contracts/api/clients/nodejs-fetch/test/protocol/pollBackoff.test.ts
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

import assert from 'node:assert/strict'
import { test } from 'node:test'
import { PollBackoff } from '../../src/index.js'

test('recovers from a hasMore burst instead of returning zero', () => {
  // hasMore sets the interval to 0 so the next poll is immediate, and 0 * multiplier is still 0 —
  // without a floor, a burst that drained left the loop polling flat out forever.
  const backoff = PollBackoff.of(1_000, 30_000, 3)
  assert.equal(backoff.afterHasMore(), 0)
  assert.equal(backoff.afterIdlePoll(0), 1_000)
  assert.equal(backoff.afterIdlePoll(1_000), 3_000)
  assert.equal(backoff.afterIdlePoll(20_000), 30_000)
  assert.equal(backoff.afterResults(), 1_000)
  assert.equal(backoff.initial(), 1_000)
})

test('a kick only ever shortens the interval', () => {
  const backoff = PollBackoff.of(1_000, 30_000, 3)
  assert.equal(backoff.afterKick(9_000, 3_000), 3_000)
  assert.equal(backoff.afterKick(1_000, 3_000), 1_000)
})

test('rejects a non-positive minimum, a maximum below it, and a multiplier of one', () => {
  assert.throws(() => PollBackoff.of(0, 30_000, 3), RangeError)
  assert.throws(() => PollBackoff.of(1_000, 500, 3), RangeError)
  assert.throws(() => PollBackoff.of(1_000, 30_000, 1), RangeError)
})
